include/nbl/builtin/hlsl/bxdf/common.hlsl

```diff
@@ -48,358 +174,738 @@ struct Basic
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((cache.getBdotH()), ::nbl::hlsl::is_same_v, typename T::scalar_type))
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((T::create(V,V)), ::nbl::hlsl::is_same_v, T))
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((T::create(V,V,b0,pNdotL,pNdotL)), ::nbl::hlsl::is_same_v, T))
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((T::createForReflection(V,V,pNdotL)), ::nbl::hlsl::is_same_v, T))
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((T::template createForReflection<surface_interactions::SAnisotropic<surface_interactions::SIsotropic<ray_dir_info::SBasic<typename T::scalar_type> > >,SLightSample<ray_dir_info::SBasic<typename T::scalar_type> > >(aniso,_sample)), ::nbl::hlsl::is_same_v, T))
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((T::create(V,V,V,V,V,eta,V)), ::nbl::hlsl::is_same_v, T))
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((T::template create<surface_interactions::SAnisotropic<surface_interactions::SIsotropic<ray_dir_info::SBasic<typename T::scalar_type> > >,SLightSample<ray_dir_info::SBasic<typename T::scalar_type> > >(aniso,_sample,eta)), ::nbl::hlsl::is_same_v, T))
+    ((NBL_CONCEPT_REQ_TYPE_ALIAS_CONCEPT)(CreatableIsotropicMicrofacetCache, typename T::isocache_type))
+);
+#undef eta
+#undef b0
+#undef V
+#undef _sample
+#undef pNdotL
+#undef aniso
+#undef cache
+#include <nbl/builtin/hlsl/concepts/__end.hlsl>
+
+template<class IsoCache NBL_PRIMARY_REQUIRES(CreatableIsotropicMicrofacetCache<IsoCache>)
+struct SAnisotropicMicrofacetCache
 {
-  // always valid by construction
-  static AnisotropicMicrofacetCache create(const float3 tangentSpaceV, const float3 tangentSpaceH)
-  {
-    AnisotropicMicrofacetCache retval;
-    
-    retval.VdotH = dot(tangentSpaceV,tangentSpaceH);
-    retval.LdotH = retval.VdotH;
-    retval.NdotH = tangentSpaceH.z;
-    retval.NdotH2 = retval.NdotH*retval.NdotH;
-    retval.TdotH = tangentSpaceH.x;
-    retval.BdotH = tangentSpaceH.y;
-    
-    return retval;
-  }
-  static AnisotropicMicrofacetCache create(
-    const float3 tangentSpaceV, 
-    const float3 tangentSpaceH,
-    const bool transmitted,
-    const float rcpOrientedEta,
-    const float rcpOrientedEta2
-  )
-  {
-    AnisotropicMicrofacetCache retval = create(tangentSpaceV,tangentSpaceH);
-    if (transmitted)
-    {
-      const float VdotH = retval.VdotH;
-      LdotH = transmitted ? refract_compute_NdotT(VdotH<0.0,VdotH*VdotH,rcpOrientedEta2);
+    using this_t = SAnisotropicMicrofacetCache<IsoCache>;
+    using isocache_type = IsoCache;
+    using scalar_type = typename IsoCache::scalar_type;
+    using vector3_type = vector<scalar_type, 3>;
+    using matrix3x3_type = matrix<scalar_type, 3, 3>;
+
+    // always valid by construction
+    static this_t create(NBL_CONST_REF_ARG(vector3_type) tangentSpaceV, NBL_CONST_REF_ARG(vector3_type) tangentSpaceH)
+    {
+        this_t retval;
+
+        retval.iso_cache.VdotH = nbl::hlsl::dot<vector3_type>(tangentSpaceV,tangentSpaceH);
+        retval.iso_cache.LdotH = retval.iso_cache.getVdotH();
```

Comment:
shouldn't this be called `createForReflection` instead of just `create` ? to keep convention with the isotropic cache methods?